Mathlib/Combinatorics/Quiver/Path/Vertices.lean

```diff
@@ -160,7 +160,325 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
     vertices_getLast p (vertices_ne_nil p)
-  have h₂ := List.getLast_mem (l := p.vertices) (vertices_ne_nil p)
+  have h₂ := getLast_mem (l := p.vertices) (vertices_ne_nil p)
   simpa [h₁] using h₂
 
+/-!  ### Path vertices decomposition -/
+
+section
+
+variable {V : Type*} [Quiver V]
+
+open List
+
+/-- Given a path `p : Path a b` and an index `n ≤ p.length`,
+    we can split `p = p₁.comp p₂` with `p₁.length = n`. -/
+theorem exists_comp_of_length_le {a b : V} (p : Path a b) {n : ℕ} (hn : n ≤ p.length) :
+    ∃ (c : V) (p₁ : Path a c) (p₂ : Path c b),
+      p = p₁.comp p₂ ∧ p₁.length = n := by
+  induction p generalizing n with
+  | nil =>
+      have h : n = 0 := by simp_all only [length_nil, nonpos_iff_eq_zero]
+      subst h
+      exact ⟨a, Path.nil, Path.nil, by simp only [comp_nil], rfl⟩
+  | cons p' e ih =>
+      rename_i c
+      rw [length_cons] at hn
+      rcases (Nat.le_succ_iff).1 hn with h | h
+      · rcases ih h with ⟨d, p₁, p₂, hp, hl⟩
+        refine ⟨d, p₁, p₂.cons e, ?_, hl⟩
+        simp; rw [hp]
+      · subst h
+        refine ⟨c, p'.cons e, Path.nil, ?_, ?_⟩
+        all_goals simp
+
+/-- If a vertex `v` occurs in the list of vertices of a path `p : Path a b`, then `p` can be
+decomposed as a concatenation of a subpath from `a` to `v` and a subpath from `v` to `b`. -/
+theorem exists_comp_of_mem_vertices {a b v : V} (p : Path a b)
+  (h : v ∈ p.vertices) : ∃ (p₁ : Path a v) (p₂ : Path v b), p = p₁.comp p₂ := by
+  obtain ⟨l₁, l₂, hv⟩ := exists_mem_split h
+  have h_len : l₁.length ≤ p.length := by
+    have : p.vertices.length = p.length + 1 := vertices_length p
+    have : l₁.length < p.vertices.length := by
+      rw [hv, length_append]
+      simp
+    omega
+  obtain ⟨c, p₁, p₂, hp, hl⟩ := exists_comp_of_length_le p h_len
+  suffices hvc : v = c by
+    subst hvc
+    exact ⟨p₁, p₂, hp⟩
+  have h_verts : p.vertices = p₁.vertices.dropLast ++ p₂.vertices := by
+    rw [hp, vertices_comp]
+  have h_l1_len : l₁.length = p₁.vertices.dropLast.length := by aesop
+  have h_l1_eq : l₁ = p₁.vertices.dropLast := by
+    have : l₁ ++ v :: l₂ = p₁.vertices.dropLast ++ p₂.vertices := by
+      rw [← hv, h_verts]
+    exact append_inj_left this h_l1_len
+  have h_v_l2 : v :: l₂ = p₂.vertices := by
+    have : l₁ ++ v :: l₂ = p₁.vertices.dropLast ++ p₂.vertices := by
+      rw [← hv, h_verts]
+    rw [h_l1_eq] at this
+    exact append_cancel_left this
+  have : p₂.vertices.head? = some c := by
+    cases p₂ with
+    | nil => simp only [vertices_nil, head?_cons]
+    | cons _ _ => exact vertices_head? _
+  rw [← h_v_l2] at this
+  simp [head?_cons, Option.some.injEq] at this
+  exact this
+
+lemma vertices_head_eq_start {a b : V} (p : Path a b) :
+    p.vertices.head (vertices_ne_nil p) = a := by
+  induction p with
+  | nil => simp only [vertices_nil, head_cons]
+  | cons p' _ ih =>
+    simp [vertices_cons, concat_eq_append]
+    have : p'.vertices ≠ [] := vertices_ne_nil p'
+    simp [head_append_of_ne_nil this]
```

Comment:
This is just `vertices_head_eq` higher up.
```suggestion
```
(in fact `vertices_head_eq` is slightly more extensible)